test_runner/src/main/kotlin/ftl/json/SavedMatrix.kt

```diff
@@ -84,55 +87,91 @@ class SavedMatrix(matrix: TestMatrix) {
-            val stepResult = GcToolResults.getStepResult(it.toolResultsStep)
-            val testTimeSeconds = stepResult.testExecutionStep?.testTiming?.testProcessDuration?.seconds ?: return
-            val testTimeout = timeoutToSeconds(it.testSpecification?.testTimeout ?: "0s")
-
-            // if overall test duration time is higher then testTimeout flank should calculate billable minutes for testTimeout
-            val timeToBill = min(testTimeSeconds, testTimeout)
-            val billableMinutes = billableMinutes(timeToBill)
+        updateFinishedMatrixData(matrix)
+    }
 
-            if (AndroidCatalog.isVirtualDevice(it.environment?.androidDevice, matrix.projectId ?: "")) {
-                billableVirtualMinutes += billableMinutes
-            } else {
-                billablePhysicalMinutes += billableMinutes
+    private fun updateFinishedMatrixData(matrix: TestMatrix) {
+        val testExecutionsData = matrix.testExecutions.createTestExecutionDataListAsync()
+        val summedTestSuiteOverviewData =
+            testExecutionsData
+                .prepareForJUnitResult()
+                .fold(TestSuiteOverviewData(0, 0, 0, 0, 0, 0.0, 0.0)) { sum, test ->
+                    sum + test.createTestSuitOverviewData()
+                }
+
+        testExecutionsData
+            .forEach {
+                val stepOutcome = GcToolResults.getExecutionResult(it.testExecution).outcome
+                updatedFinishedInfo(
+                    stepOutcome = stepOutcome,
+                    flakyOutcome = stepOutcome.summary != it.step.outcome.summary,
+                    testSuiteOverviewData = summedTestSuiteOverviewData,
+                    isRoboTests = it.testExecution.testSpecification?.androidRoboTest != null,
+                    isVirtualDevice = isVirtualDevice(
+                        device = it.testExecution.environment.androidDevice,
+                        projectId = matrix.projectId.orEmpty()
+                    ),
+                    billableMinutes = it.testExecution.getBillableMinutes()
+                )
             }
-        }
     }
 
-    private fun updateOutcome(stepOutcome: Outcome) {
-        // the matrix outcome is failure if any step fails
-        // if the matrix outcome is already set to failure then we can ignore the other step outcomes.
-        // inconclusive is treated as a failure
-        if (outcome == failure || outcome == inconclusive) return
+    private fun TestExecution.getBillableMinutes() =
+        takeIf { testExecution -> testExecution.state != ERROR }
+            ?.run {
+                // testExecutionStep, testTiming, etc. can all be null.
+                // sometimes testExecutionStep is present and testTiming is null
+                val testTimeSeconds =
+                    GcToolResults.getStepResult(toolResultsStep).testExecutionStep?.testTiming?.testProcessDuration?.seconds
+                        ?: return@run null
```

Comment:
I think `@run` is redundant (but correct me if I'm wrong)

Reply:
100% right